numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -309,6 +308,53 @@ static int
 
 /**end repeat**/
 
+static NPY_INLINE npy_longdouble
+string_to_long_double(PyObject*op)
+{
+    char *s;
+    char *end;
+    npy_longdouble temp;
+    PyObject* b;
+
+    if (PyUnicode_Check(op)) {
```

Comment:
Checking here (as opposed to a duck-typed conversion) saves a string->unicode->string conversion on python 2. But given the error handling and the need to convert to a C string, it's easier to just convert the bytes to a C string and fall back to python conversion if that fails. EAFAP.